### Import required dependencies

In [1]:
import pandas as pd
import os

## Deliverable 1: Collect the Data

To collect the data that you’ll need, complete the following steps:

1. Using the Pandas `read_csv` function and the `os` module, import the data from the `new_full_student_data.csv` file, and create a DataFrame called student_df. 

2. Use the head function to confirm that Pandas properly imported the data.


In [2]:
# Create the path and import the data
full_student_data = os.path.join('Resources/new_full_student_data.csv')
student_df = pd.read_csv(full_student_data)

In [3]:
# Verify that the data was properly imported
student_df.head()

student_id      student_name grade           school_name  reading_score  \
0   103880842     Travis Martin   9th  Sullivan High School           59.0   
1    45069750     Michael Brown   9th     Dixon High School           94.7   
2    45024902   Gabriela Lucero   9th    Wagner High School           89.0   
3    62582498  Susan Richardson   9th     Silva High School           69.7   
4    16437227      Sherry Davis  11th    Bowers High School            NaN   

   math_score school_type  school_budget  
0        88.2      Public         961125  
1        73.5     Charter         870334  
2        70.4      Public         846745  
3        80.3      Public         991918  
4        27.5      Public         848324

## Deliverable 2: Prepare the Data

To prepare and clean your data for analysis, complete the following steps:
    
1. Check for and remove all rows with `NaN`, or missing, values in the student DataFrame. 

2. Check for and remove all duplicate rows in the student DataFrame.

3. Use the `str.replace` function to remove the "th" from the grade levels in the grade column.

4. Check data types using the `dtypes` property.

5. Remove the "th" suffix from every value in the grade column using `str` and `replace`.

6. Change the grade colum to the `int` type and verify column types.

7. Use the head (and/or the tail) function to preview the DataFrame.

In [4]:
# Check for null values
student_df.isna().sum()

student_id          0
student_name        0
grade               0
school_name         0
reading_score    1968
math_score        982
school_type         0
school_budget       0
dtype: int64

In [5]:
# Drop rows with null values and verify removal
student_df = student_df.dropna()
student_df.isna().sum()

student_id       0
student_name     0
grade            0
school_name      0
reading_score    0
math_score       0
school_type      0
school_budget    0
dtype: int64

In [6]:
# Check for duplicated rows
student_df.duplicated().sum()

1836

In [7]:
# Drop duplicated rows and verify removal
student_df=student_df.drop_duplicates()
student_df.duplicated().sum()

0

In [8]:
# Check data types
student_df.dtypes

student_id         int64
student_name      object
grade             object
school_name       object
reading_score    float64
math_score       float64
school_type       object
school_budget      int64
dtype: object

In [9]:
# Examine the grade column to understand why it is not an int
student_df['grade']

0         9th
1         9th
2         9th
3         9th
5         9th
         ... 
19508    10th
19509    12th
19511    11th
19512    11th
19513    12th
Name: grade, Length: 14831, dtype: object

In [10]:
# Remove the non-numeric characters and verify the contents of the column
student_df['grade'] = student_df['grade'].str.replace('th', '')
student_df['grade']

0         9
1         9
2         9
3         9
5         9
         ..
19508    10
19509    12
19511    11
19512    11
19513    12
Name: grade, Length: 14831, dtype: object

In [11]:
# Change the grade column to the int type and verify column types
student_df['grade'] = student_df['grade'].astype(int)
student_df.dtypes

student_id         int64
student_name      object
grade              int32
school_name       object
reading_score    float64
math_score       float64
school_type       object
school_budget      int64
dtype: object

## Deliverable 3: Summarize the Data

Describe the data using summary statistics on the data as a whole and on individual columns.

1. Generate the summary statistics for each DataFrame by using the `describe` function.

2. Display the mean math score using the `mean` function. 

2. Store the minimum reading score as `min_reading_score`.

In [12]:
# Display summary statistics for the DataFrame
student_df.describe()

student_id         grade  reading_score    math_score  school_budget
count  1.483100e+04  14831.000000   14831.000000  14831.000000   14831.000000
mean   6.975296e+07     10.355539      72.357865     64.675733  893742.749107
std    3.452909e+07      1.097728      15.224590     15.844093   53938.066467
min    1.000906e+07      9.000000      10.500000      3.700000  817615.000000
25%    3.984433e+07      9.000000      62.200000     54.500000  846745.000000
50%    6.965978e+07     10.000000      73.800000     65.300000  893368.000000
75%    9.927449e+07     11.000000      84.000000     76.000000  956438.000000
max    1.299997e+08     12.000000     100.000000    100.000000  991918.000000

In [13]:
# Display the mean math score using the mean function
student_df["math_score"].mean()

64.67573326141189

In [14]:
# Store the minimum reading score as min_reading_score
min_reading_score = student_df['reading_score'].min()
min_reading_score

10.5

In [15]:
student_df['reading_score'].mean()

72.35786528217913

## Deliverable 4: Drill Down into the Data

Drill down to specific rows, columns, and subsets of the data.

To drill down into the data, complete the following steps:

1. Use `loc` to display the grade column.

2. Use `iloc` to display the first 3 rows and columns 3, 4, and 5.

3. Show the rows for grade nine using `loc`.

4. Store the row with the minimum overall reading score as `min_reading_row` using `loc` and the `min_reading_score` found in Deliverable 3.

5. Find the reading scores for the school and grade from the output of step three using `loc` with multiple conditional statements.

6. Using conditional statements and `loc` or `iloc`, find the mean reading score for all students in grades 11 and 12 combined.

In [16]:
# Use loc to display the grade column
#student_df['grade']
student_df.iloc[:,2]

0         9
1         9
2         9
3         9
5         9
         ..
19508    10
19509    12
19511    11
19512    11
19513    12
Name: grade, Length: 14831, dtype: int32

In [17]:
# Use `iloc` to display the first 3 rows and columns 3, 4, and 5.
student_df.iloc[0:3, 3:6]

school_name  reading_score  math_score
0  Sullivan High School           59.0        88.2
1     Dixon High School           94.7        73.5
2    Wagner High School           89.0        70.4

In [18]:
# Select the rows for grade nine and display their summary statistics using `loc` and `describe`.
grade9_df = student_df.loc[student_df["grade"] == 9]
grade9_df.describe()

student_id   grade  reading_score   math_score  school_budget
count  4.132000e+03  4132.0    4132.000000  4132.000000    4132.000000
mean   6.979441e+07     9.0      69.236713    66.585624  898692.606002
std    3.470565e+07     0.0      15.277354    16.661533   54891.596611
min    1.000906e+07     9.0      17.900000     5.300000  817615.000000
25%    3.953848e+07     9.0      59.000000    56.000000  846745.000000
50%    6.984037e+07     9.0      70.050000    67.800000  893368.000000
75%    9.939504e+07     9.0      80.500000    78.500000  957299.000000
max    1.299997e+08     9.0      99.900000   100.000000  991918.000000

In [19]:
# Store the row with the minimum overall reading score as `min_reading_row`
# using `loc` and the `min_reading_score` found in Deliverable 3.
min_reading_row = student_df.loc[student_df["reading_score"] == min_reading_score]
min_reading_row

student_id    student_name  grade        school_name  reading_score  \
3706    81758630  Matthew Thomas     10  Dixon High School           10.5   

      math_score school_type  school_budget  
3706        58.4     Charter         870334

In [20]:
# Use loc with conditionals to select all reading scores from 10th graders at Dixon High School

sortedList_df = student_df.loc[(student_df["school_name"] == "Dixon High School") & (student_df["grade"] == 10)]
sortedList_df.loc[:, ["school_name", "reading_score"]]


school_name  reading_score
45     Dixon High School           71.1
60     Dixon High School           59.5
69     Dixon High School           88.6
94     Dixon High School           81.5
100    Dixon High School           95.3
...                  ...            ...
19283  Dixon High School           52.9
19306  Dixon High School           58.0
19344  Dixon High School           38.0
19368  Dixon High School           84.4
19445  Dixon High School           43.9

[569 rows x 2 columns]

In [34]:
# Find the mean reading score for all students in grades 11 and 12 combined.

grade_df = student_df.loc[(student_df["grade"] >10) & (student_df["grade"] <13)]
mean_reading = grade_df["reading_score"].mean()

mean_reading


74.90038089192188

## Deliverable 5: Make Comparisons Between District and Charter Schools
(From starter file - does not match Online Challenge)

Compare district vs charter schools for budget, size, and scores.

Make comparisons within your data by completing the following steps:

1. Using the `groupby` and `mean` functions, look at the average reading and math scores per school type.

1. Using the `groupby` and `count` functions, find the total number of students at each school.

2. Using the `groupby` and `mean` functions, find the average budget per grade for each school type.

In [22]:
# Use groupby and mean to find the average reading and math scores for each school type.

avg_reading_by_type = student_df.groupby(by="school_type").mean()
avg_reading_by_type.loc[:, ["reading_score", "math_score"]]

reading_score  math_score
school_type                           
Charter          72.450603   66.761883
Public           72.281219   62.951576

In [35]:
# Use the `groupby`, `count`, and `sort_values` functions to find the
# total number of students at each school and sort from most students to least students.
avg_reading_by_type = student_df.groupby(by="school_name").mean()

In [24]:
student_by_school = student_df.groupby(by="school_name").sum()
student_by_school

student_id  grade  reading_score  math_score  \
school_name                                                              
Bowers High School       55635515909   8925        61228.8     42710.8   
Campbell High School     28445373550   4264        28954.9     25306.9   
Campos High School       37179626255   5312        36745.2     32112.4   
Chang High School        12502094414   1723        11401.6     10672.5   
Dixon High School       111213157589  16122       117312.5    109393.9   
Fisher High School       56392126853   8379        57203.2     52667.3   
Green High School       137073024519  19679       148384.6    129865.6   
Montgomery High School  142472162238  21889       142711.9    136163.9   
Odonnell High School     31743370661   4974        31719.2     31117.2   
Richard High School      38334560731   5760        37450.5     36636.2   
Silva High School        77874765313  11461        81308.6     67307.7   
Sullivan High School     66218240554   9592        69734.3     59739.8   
Turner High School       59108239158   8316        64172.5     55873.4   
Wagner High School      108474692154  16154       113338.9    100107.8   
Woods High School        71839227267  11033        71472.8     69530.4   

                        school_budget  
school_name                            
Bowers High School          681204172  
Campbell High School        391015482  
Campos High School          517898759  
Chang High School           143662572  
Dixon High School          1377738722  
Fisher High School          652456770  
Green High School          1632865870  
Montgomery High School     1820683984  
Odonnell High School        440398566  
Richard High School         490224149  
Silva High School          1100037062  
Sullivan High School        933252375  
Turner High School          809146548  
Wagner High School         1304834045  
Woods High School           959679636

## Deliverable 5: Compare the Data
1. Display the average budget for each school type by using the groupby and mean functions, as the following image shows:


In [25]:
budget_by_type = student_df.groupby(by="school_type").mean()
budget_by_type.loc[:,["school_budget"]]


school_budget
school_type               
Charter      872625.656236
Public       911195.558251

  2. Find the total number of students at each school, and sort those numbers from largest to smallest by using the groupby, count, and sort_values functions, as the following image shows:

In [26]:
group_schools_df = student_df.groupby(["school_name"])["student_name"].count().reset_index(name="student_count")

sorted_group_df = group_schools_df.sort_values("student_count", ascending=False)

sorted_group_df

school_name  student_count
7   Montgomery High School           2038
6        Green High School           1961
4        Dixon High School           1583
13      Wagner High School           1541
10       Silva High School           1109
14       Woods High School           1052
11    Sullivan High School            971
12      Turner High School            846
0       Bowers High School            803
5       Fisher High School            798
9      Richard High School            551
2       Campos High School            541
8     Odonnell High School            459
1     Campbell High School            407
3        Chang High School            171

3. Find the average math score by grade for each school type by using the groupby and mean functions, as the following image shows:

In [27]:
avg_math_grade_school = student_df.groupby(["school_type", "grade"]).mean()

avg_math_grade_school.loc[:,["math_score"]].round(0)

math_score
school_type grade            
Charter     9            70.0
            10           66.0
            11           68.0
            12           60.0
Public      9            64.0
            10           64.0
            11           59.0
            12           64.0

# Deliverable 6: Summarize Your Findings
In the cell below, write a few sentences to describe any discoveries you made while performing your analysis along with any additional analysis you believe would be worthwhile.

## Analysis Summary
### Background
Starting with the student records from all the schools in the district, the school board requested that we provide some analysis on school budgets, students per school and grades. 

### Summary
In the analysis provided the following inforamtion was produced:
* Budget: The Public school budgets were higher than the Charter schools
  * Public: 911,196 vs Charter: 872,626
* Reading & Math: Both Math and Reading scores were higher in the Charter schools vs the Public schools
  * Charter: 72.5 and 66.7 vs Public: 72.3 and 62.9
  
Overall, although there is a difference in budget the grad averages between Charter and Public schools are quite similar.

### Opportunities for further analysis

In reviewing math score by grade in Charter schools we see something of interest that could be an opportunity for further analysis.  Math scores in the Charter schools were higher than Public schools in all grades except grade 12.  In the Charter schools the math scores dropped below the Public schools in grade 12.  A further anlysis of total number of students in grade 12 may help provide forther inforamtion as to why Charter math scores dropped in the grade 12 students.

Further analysis could be performed to compary school populations vs grades as well.
